src/ecmascript_simd.js

```diff
@@ -117,7 +117,14 @@ function simdCheckLaneIndex(index, lanes) {
 var lanes = [];
 
 function simdCreate(type) {
-  return type.fn.apply(type.fn, lanes);
+  // Work around v8 NaN canonicalization issue: if lanes contains floats with non-canonical NaN bit patterns,
+  // type.fn.apply() will canonicalize the NaNs and the bits are lost (most likely as part of float->double expansion).
+  // Directly passing the arguments into the function preserves them.
+  if (type.name == "Float32x4") {
+    return SIMD.Float32x4(lanes[0], lanes[1], lanes[2], lanes[3]);
+  } else {
+    return type.fn.apply(type.fn, lanes);
+  }
```

Comment:
Yes, this change is to fix https://github.com/tc39/ecmascript_simd/issues/277 , however I don't know what the status of a fix in upstream is. Converting the whole polyfill to start using typed arrays instead of regular arrays is a larger body of work, and I don't have the capacity to start developing a pull request to fix https://github.com/tc39/ecmascript_simd/issues/277 in a way proposed there. For Emscripten test suite purposes we only need the Float32x4 to be hotpatched in the two places. I can mark these changes as Emscripten-specific with `XXX Emscripten` if that helps. Would that be ok?